### ISSUE TITLE
distribution-restrictions, numcopies, and required content

### ISSUE THREAD

**Issue by kyleam** [datalad/datalad#4605]
The metadata section of the manual [describes](https://docs.datalad.org/en/stable/metadata.html#automating-non-distribution-of-sensitive-files) using the `distribution-restrictions` field to mark files that shouldn't be shared publicly. This is a dedicated issue to track the concern I've mentioned elsewhere. In order to expect `datalad publish` to not transfer marked data, callers need to worry about and monitor the `numcopies` setting.

<details><summary>example</summary>

```sh
#!/bin/sh

cd "$(mktemp -d ${TMPDIR:-/tmp}/dl-XXXXXXX)"

datalad create a
(
    cd a
    git config receive.denyCurrentBranch updateInstead
    git annex numcopies 2
)

datalad clone a b
(
    cd b
    echo foo >foo
    echo bar >bar
    datalad save
    git annex metadata --set distribution-restrictions=sensitive bar
    git annex wanted origin "not metadata=distribution-restrictions=*"
    datalad publish --to origin
)


(
    cd a
    cat foo
    cat bar
)
```

</details>

This setting is usually shared between clones, so it is also possible for users that are aware of the `numcopies` behavior to accidentally publish sensitive information because they didn't catch an incoming change in the `numcopies` value.

Updating `publish` to override `numcopies` in its call to `git annex copy` would be sufficient to prevent the above case, I think. Another option would be to try to detect whether `numcopies` is set above 1 and abort if `--auto` will be used. Yet another possibility would be to update the documentation to suggest setting a file-level `annex.numcopies` value in .gitattributes.

When thinking about how to address this, it is worth considering that very similar concerns apply to required content. (As an example, replace the `git annex numcopies` call in the above script with `git annex required . anything`.)

In my view the minimum action needed is an update of the docs to provide a warning that the suggested approach is dangerous if `numcopies` is above the default value of 1 or if required content is configured.

**Comment by yarikoptic** [comment]
> Another option would be to try to detect whether `numcopies` is set above 1 and abort if `--auto` will be used.

I think this is the best course of action ATM for both `publish` and `push` if that one would anyhow enable `--auto`.

> Yet another possibility would be to update the documentation to suggest setting a file-level `annex.numcopies` value in .gitattributes.

would it take precedence over "global" numcopies? (sorry - have not tried yet fully; at least even with it in .gitattributes to 1 I still get `git annex numcopies` reporting 2)

**Comment by yarikoptic** [comment]
And instead of global "numcopies" may be it is feasible to use `lackingcopies=` in wanted expressions so to fulfill "numcopies" desire while warranting restrictions of "wanted"? If so (didn't try yet) - we could then recommend that in the docs for wherever needed.

PS Sure thing we could also add a special treatment to detecting "distribution-restrictions" in "wanted", but I think that would be too ad-hoc/hardcoded

**Comment by mih** [comment]
> Updating publish to override numcopies in its call to git annex copy would be sufficient to prevent the above case, I think.

I confirm and agree.

<details><summary>Demo that `copy -N 1` takes precedence over numcopies from .gitattributes</summary>

```sh
#!/bin/sh

cd "$(mktemp -d ${TMPDIR:-/tmp}/dl-XXXXXXX)"

datalad create a
(
    cd a
    git config receive.denyCurrentBranch updateInstead
    git annex numcopies 2
    echo "bar annex.numcopies=2" > .gitattributes && git add .gitattributes && git commit -m attr
)

datalad clone a b
(
    cd b
    echo foo >foo
    echo bar >bar
    datalad save
    git annex metadata --set distribution-restrictions=sensitive bar
    git annex wanted origin "not metadata=distribution-restrictions=*"
    git annex copy -N 1 --to origin --auto
)


(
    cd b
    echo "Must not show 'bar'"
    git annex find --in origin
)
```
</details>

> Another option would be to try to detect whether numcopies is set above 1 and abort if --auto will be used.

This would be a very expensive operation. In order to be valid, this check needs to be performed per file, because file-based `numcopies` settings can come in from any location that can provide `gitattributes`.

> Yet another possibility would be to update the documentation to suggest setting a file-level annex.numcopies value in .gitattributes.

This would enable "remote" lifting of distribution restrictions via fetch/merge. This may not be a bad thing, because the distribution restrictions are defined in the same mutable fashion.

### My conclusion

Overriding `numcopies` disables a legitimately useful mechanism. Instead, providing 
```
* annex.numcopies=1
```
as a top-level dataset gitattribute, to be added by a maintainer who wants `copy --auto` to work reliably with a preferred content setup that honors distribution restriction conventions, is minimal extra effort (`wanted` expression has to be set up anyways), and does not conflict with other use cases.

My vote is on the documentation change, possibly in conjuction with a config procedure `cfg_distribution-restrictions` that established the complete setup.

**Comment by kyleam** [comment]
> My vote is on the documentation change, possibly in conjuction with a config procedure `cfg_distribution-restrictions` that established the complete setup.

I agree with your assessment, though I'm pretty lukewarm on the idea of the procedure.  I think it is probably a good thing that distributions-restrictions approach isn't mentioned aside from one spot in the manual, and even there it's a terse "we did this" sort of description.  The approach depends on git-annex details that I think users should research and think through beforehand.

Anyway, as a first step at least, I've added a warning in gh-4615.
